source/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,351 @@
+	Exclusive        bool   `envconfig:"RABBITMQ_QUEUE_CONFIG_EXCLUSIVE" required:"false"`
+	NoWait           bool   `envconfig:"RABBITMQ_QUEUE_CONFIG_NOWAIT" required:"false"`
+}
+
+type adapterConfig struct {
+	adapter.EnvConfig
+
+	Brokers        string `envconfig:"RABBITMQ_BROKERS" required:"true"`
+	Topic          string `envconfig:"RABBITMQ_TOPIC" required:"true"`
+	User           string `envconfig:"RABBITMQ_USER" required:"false"`
+	Password       string `envconfig:"RABBITMQ_PASSWORD" required:"false"`
+	ChannelConfig  ChannelConfig
+	ExchangeConfig ExchangeConfig
+	QueueConfig    QueueConfig
+}
+
+func NewEnvConfig() adapter.EnvConfigAccessor {
+	return &adapterConfig{}
+}
+
+type Adapter struct {
+	config 			  *adapterConfig
+	httpMessageSender *kncloudevents.HttpMessageSender
+	reporter 		  source.StatsReporter
+	logger 			  *zap.Logger
+	context			  context.Context
+}
+
+var _ adapter.MessageAdapter = (*Adapter)(nil)
+var _ adapter.MessageAdapterConstructor = NewAdapter
+
+func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, httpMessageSender *kncloudevents.HttpMessageSender, reporter source.StatsReporter) adapter.MessageAdapter {
+	logger := logging.FromContext(ctx).Desugar()
+	config := processed.(*adapterConfig)
+
+	return &Adapter{
+		config: 		   config,
+		httpMessageSender: httpMessageSender,
+		reporter:		   reporter,
+		logger: 		   logger,
+		context: 		   ctx,
+	}
+}
+
+func (a *Adapter) CreateConn(User string, Password string, logger *zap.Logger) (*amqp.Conn, error) {
+	if User != "" && Password != "" {
+		a.config.Brokers = fmt.Sprintf("amqp://%s:%s@%s", a.config.User, a.config.Password, a.config.Brokers)
+	}
+	conn, err := amqp.Dial(a.config.Brokers)
+	if err != nil {
+		logger.Error(err.Error())
+	}
+	return conn, err
+}
+
+func (a *Adapter) CreateChannel(conn *amqp.Conn, connTest *amqptest.Conn,
+	logger *zap.Logger) (wabbit.Channel, error) {
+	var ch wabbit.Channel
+	var err error
+
+	if conn != nil {
+		ch, err = conn.Channel()
+	} else {
+		ch, err = connTest.Channel()
+	}
+	if err != nil {
+		logger.Error(err.Error())
+		return nil, err
+	}
+
+	err = ch.Qos(a.config.ChannelConfig.PrefetchCount,
+		0,
+		a.config.ChannelConfig.GlobalQos)
+
+	return ch, err
+}
+
+func (a *Adapter) Start(stopCh <-chan struct{}) error {
+	logger := a.logger
+
+	logger.Info("Starting with config: ", zap.Any("adapter", a))
+
+	conn, err := a.CreateConn(a.config.User, a.config.Password, logger)
+	if err == nil {
+		defer conn.Close()
+	}
+
+	ch, err := a.CreateChannel(conn, nil, logger)
+	if err == nil {
+		defer ch.Close()
+	}
+
+	queue, err := a.StartAmqpClient(&ch)
+	if err != nil {
+		logger.Error(err.Error())
+	}
+
+	return a.PollForMessages(&ch, queue, stopCh)
+}
+
+func (a *Adapter) StartAmqpClient(ch *wabbit.Channel) (*wabbit.Queue, error) {
+	logger := a.logger
+	exchangeConfig := fillDefaultValuesForExchangeConfig(&a.config.ExchangeConfig, a.config.Topic)
+
+	err := (*ch).ExchangeDeclare(
```

Comment:
Perhaps the lifecycle of these resources (Exchange, Queue and the Bindings below) should be managed by the reconciler? That would allow for updates/deletes to occur there as well. This code could then be changed to use "passive" versions of the declare operations.

As far as I can tell, in the current implementation, even if a Source resource is deleted, these resources will remain (which means the Queue continues to receive messages via the Binding).

Reply:
Cool, thanks 👍